### PR TITLE
Rework implant surgery to be plausibly deniable/deliberate

### PR DESCRIPTION
**Pull request by coiax**
:cl: coiax
balance: Implant removal surgery has been reworked into implant extraction/removal surgery.
add: After the scalpel, hemostat, retractor, you enter the "searching" step, where you need to fish around in the person's chest with either your hands or a t-ray scanner. If you don't find anything, that could mean there are no implants, or you just didn't find it. You can look as many times as you want. (A t-ray scanner has a higher chance of finding an implant, if it exists, than just using your bare hands.) add: Once you've found an implant, you either extract it with a hemostat, or destroy it with a surgical drill. You are not allowed to use extraction tools without an implant case nearby that can recieve the implant, use destruction tools instead.
add: When using extraction/destruction tools, you get to choose which implant you're targeting, just like manipulate organs. No more will you destroy the clown's prize sad trombone implant while trying to prepare them for revolutionary fervour.
/:cl:

This is motivated by three distinct goals
- To give traitors with bad implants plausible deniability if exploratory surgery doesn't find anything (because you could have just been unlucky), and thus make it uneconomical to just cut open someone up on a tiny suspicion.

- The destruction/extraction of an implant is now a deliberate choice, and not just something you need to go "oh well let's hope the game recognises there's an implantcase nearby" if you're intending to keep it.

Assuming security love gamer traitor loot, they'll likely be more keen to have implant cases nearby when taking a traitors uplink or storage implant out.

Revolutionaries de-mindshielding crew however don't care about salvaging the implant, and will be using the drill or screwdriver instead.

- The implant you're targetting is now via a list, so you don't accidentally remove the clown's sad trombone implant when you're trying to remove their mindshield for revolutionary purposes.

Also, clarifies some out of date comments on the surgery step code.

### CODE REVIEW

**Comment by Jacquerel** [comment]
I like this concept, but wonder if it would be possible to include the medical scanner as well as the T-Ray scanner (it can still be worse at the job than the T-Ray scanner) as a potential tool to use here, as you would hope it would usually be medbay performing these operations and T-Ray scanners are generally not owned by anyone outside of Engineering.

**Comment by coiax** [comment]
> I like this concept, but wonder if it would be possible to include the medical scanner as well as the T-Ray scanner (it can still be worse at the job than the T-Ray scanner) as a potential tool to use here, as you would hope it would usually be medbay performing these operations and T-Ray scanners are generally not owned by anyone outside of Engineering.

As in the medical analyzer? It didn't really occur to me as an option, but it's a good idea. Maybe with the advanced medical scanner being better at it?

**Comment by Jacquerel** [comment]
That sounds good to me.

File: code/modules/surgery/implant_removal.dm [hunk 1]
```diff
@@ -1,81 +1,199 @@
...
 	implements = list(
-		TOOL_HEMOSTAT = 100,
-		TOOL_CROWBAR = 65,
-		/obj/item/kitchen/fork = 35)
-	time = 64
-	success_sound = 'sound/surgery/hemostat1.ogg'
-	var/obj/item/implant/implant
+		/obj/item/t_scanner = 100,
+	)
+	time = 12.8 SECONDS
```
Comment (Jacquerel) [inline code/modules/surgery/implant_removal.dm:20]: This seems like a _very_ long time, especially for a repeatable step with a reasonably high failure chance which may need to be performed multiple times.
Making implants more deniable is a good idea but I think that even people it is being done to won't appreciate being on the table for the additional _minutes_ it takes to have a reasonable chance of discovering an implant using the best possible tool.

File: code/modules/surgery/implant_removal.dm [hunk 2]
```diff
@@ -1,81 +1,199 @@
...
+	repeatable = TRUE
+	var/detection_prob = list(
```
Comment (Jacquerel) [inline code/modules/surgery/implant_removal.dm:22]: New vars need documenting comments, true in general across this file.

File: code/modules/surgery/implant_removal.dm [hunk 3]
```diff
@@ -1,81 +1,199 @@
...
+				span_notice("You begin to extract [chosen_implant] from [target]'s [parse_zone(target_zone)]..."),
+				span_notice("[user] begins to extract [chosen_implant] from [target]'s [parse_zone(target_zone)]."),
+				span_notice("[user] begins to extract something from [target]'s [parse_zone(target_zone)]."),
+			)
+			// I'm sorry, but you cannot tell what type of implants is being removed from sensation alone.
+			display_pain(target, "You can feel an implant being removed from your [parse_zone(target_zone)]!")
+		else
+			return SURGERY_STEP_FAIL
+
+	else
```
Comment (Jacquerel) [inline code/modules/surgery/implant_removal.dm:143]: Does this need to be an else, or can you just ensure that the previous `if` always returns eventually?

File: code/modules/surgery/implant_removal.dm [hunk 4]
```diff
@@ -1,81 +1,199 @@
...
 			)
+			// I'm sorry, but you cannot tell what type of implants is being removed from sensation alone.
+			display_pain(target, "You can feel an implant being destroyed in your [parse_zone(target_zone)]!")
 		else
-			qdel(implant)
+			return SURGERY_STEP_FAIL
 
+/datum/surgery_step/destroy_extract_implant/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/implant_removal/surgery, default_display_results = FALSE)
+	if(!target_implant || !(target_implant in target.implants))
+		to_chat(user, span_warning("You can't find what you were looking for in [target]'s [target_zone]!"))
```
Comment (Jacquerel) [inline code/modules/surgery/implant_removal.dm:164]:
```suggestion
	if(!target_implant || !(target_implant in target.implants))
		to_chat(user, span_warning("You can't find what you were looking for in [target]'s [target_zone]!"))
		return ..()
```

It feels like you can use early returns pretty extensively in this proc to reduce how far everything below is indented, might make the execution a little clearer.

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Couple of nitpicks, mostly that you have a lot of `if/else` trees which could be streamlined by just returning earlier.

**Comment by VladinXXV** [comment]
> * The destruction/extraction of an implant is now a deliberate choice, and not just something you need to go "oh well let's hope the game recognises there's an implantcase nearby" if you're intending to keep it.

The game doesn't recognize an implant case "nearby" because it isn't supposed to. You need to hold it in your off-hand. Even the wiki gets this right.
image
It has **never** failed on me. Not once.
> Assuming security love gamer traitor loot, they'll likely be more keen to have implant cases nearby when taking a traitors uplink or storage implant out.

Uplink implants do not work on re-implant. I actually went ahead and tested all the traitor/nukie implants just now - uplink and apparently deathrattle are the only ones that do not work on re-implant (at least, into nonantags. i didnt bother with antags.) I will also mention that all nine implants were originally implanted into a traitor and then extracted from said traitor using implant removal surgery as I described. No issues there.

File: code/modules/surgery/implant_removal.dm [hunk 5]
```diff
@@ -1,81 +1,199 @@
...
+		/datum/surgery_step/destroy_extract_implant,
 		/datum/surgery_step/close,
 	)
+	var/list/discovered_implants = list()
 
-//extract implant
-/datum/surgery_step/extract_implant
-	name = "extract implant (hemostat)"
+/datum/surgery_step/search_for_implants
+	name = "search for implants (hand/t-ray)"
```
Comment (VladinXXV) [inline code/modules/surgery/implant_removal.dm:15]: This doesn't mention health analyzers at all. I would make it say (health analyzer/hand)

File: code/modules/surgery/implant_removal.dm [hunk 6]
```diff
@@ -1,81 +1,199 @@
...
+		/obj/item/t_scanner = 100,
+	)
+	time = 12.8 SECONDS
+	repeatable = TRUE
+	var/detection_prob = list(
+		/obj/item/t_scanner = 20,
+		/obj/item/healthanalyzer/advanced = 15,
+		/obj/item/healthanalyzer = 10,
+		"" = 5, // empty string is no tool, so hand.
+	)
```
Comment (VladinXXV) [inline code/modules/surgery/implant_removal.dm:27]: Having a roundstart engineering tool be better than head of staff/researched medical gear? That's just silly. Construction tools are generally the ghetto option and I think it should stay that way. Health analyzers are professional medical tools, the t-ray is for looking at pipes. You should make the ranking for this:

1. Advanced Health Analyzer = 20
2. Health Analyzer = 15
3. T-Ray Scanner = 10
4. Empty Hand = 5

**Review by VladinXXV — CHANGES_REQUESTED** [review]
thanks for tricking me into testing what implants work when re-implanted, i found a neat (existing, not your fault) bug i can report now.

File: code/modules/surgery/implant_removal.dm (hunk 1 above)
Comment (coiax) [reply code/modules/surgery/implant_removal.dm:20]: I figured that the numbers would need tweaking, do you have better suggestions for detection prob/time for step?

Comment (VladinXXV) [reply code/modules/surgery/implant_removal.dm:20]: I'm thinking 5 seconds would be good, but Jacq may have something else in mind.

File: code/modules/surgery/implant_removal.dm (hunk 6 above)
Comment (Fikou) [reply code/modules/surgery/implant_removal.dm:27]: not a code review

Comment (VladinXXV) [reply code/modules/surgery/implant_removal.dm:27]: Seems as valid as a review complaining about how long something takes, but ok.

**Comment by Vect0r2** [comment]
Is there a way that if somebody does this with bare hands, it might infect somebody with the flu, or some other virus?

**Comment by ATH1909** [comment]
> You are not allowed to use extraction tools without an implant case nearby that can recieve the implant, use destruction tools instead.

ehhhh, this seems really game-y to me. surely there's a way to let the player signal that they're okay with destroying the implant using extraction tools?

File: code/modules/surgery/implant_removal.dm [hunk 7]
```diff
@@ -1,81 +1,199 @@
...
+/datum/surgery_step/destroy_extract_implant
+	name = "destroy/extract implant (drill/hemostat)"
+	repeatable = TRUE
+
+	// Destruction tools
+	implements = list(
+		TOOL_DRILL = 100,
+		/obj/item/screwdriver/power = 80,
+		/obj/item/pickaxe/drill = 60,
+		TOOL_SCREWDRIVER = 25,
```
Comment (ATH1909) [inline code/modules/surgery/implant_removal.dm:84]:
```suggestion
		TOOL_SCREWDRIVER = 25,
		/obj/item/pen = 10,
```

**Comment by necromanceranne** [comment]
Gonna echo vladin pretty much the whole way through. Though I am happy you're giving more uses for the drill

**Comment by iwishforducks** [comment]
lol just remove the surgery at this point if we're going through these hoops. you underestimate people's willingness to repeat this surgery 1000 times and if you don't want them doing that then just remove the surgery

**Comment by necromanceranne** [comment]
Honestly if they want to do a surgery on repeat that's their prerogative.

**Comment by RikuTheKiller** [comment]
I'd prefer if the tool just changed how quickly the surgery is done, similarly to like... idk... every other surgery in the game?

And I also agree with Vladin on the tool changes. T-ray should be ghetto, period.

**Comment by AlamoTurtle** [comment]
The only thing I could recommend here is either using a drill to destroy the implant, which is nice, or put a use to the rightclick feature of guarantee failing a surgery, so you can choose to have a menu to destroy an implant on rightclick with hemostat, which would guarantee fail the surgery and destroy the implant in the process. If you can implement destroying the implant on failed surgery, it'd be optional for hemostat even without explicitly stating as such, which I'd prefer greatly.

**Pull request closed without merging**